radix-engine/src/engine/process.rs

```diff
@@ -1005,32 +988,47 @@ impl<'r, 'l, L: SubstateStore> Process<'r, 'l, L> {
                 }
             }
             SNodeState::Bucket(bucket) => match function.as_str() {
-                "burn" => bucket.drop(self),
+                "burn" => bucket.drop(self).map_err(RuntimeError::BucketError),
                 _ => Err(RuntimeError::IllegalSystemCall),
             },
             _ => {
                 // start a new process
-                let mut process = Process::new(self.depth + 1, self.trace, self.track);
+                let mut process = Process::new(self.depth + 1, self.trace, self.track, None);
                 process.caller_auth_zone = &self.auth_zone;
 
                 // move buckets and proofs to the new process.
                 process.receive_buckets(moving_buckets)?;
-                process.receive_proofs(moving_proofs)?;
+                process.receive_proofs(moving_proofs);
 
                 // invoke the main function
-                let result = process.run(snode, function, args)?;
-
-                // figure out what buckets and resources to move from the new process
-                let moving_buckets = process.send_buckets(&result.bucket_ids)?;
-                let moving_proofs = process.send_proofs(&result.proof_ids, MoveMethod::AsReturn)?;
-
-                // drop proofs and check resource leak
-                process.drop_all_proofs()?;
-                process.check_resource()?;
+                let (result, moving_buckets, moving_proofs) =
+                    process.run(&mut snode, function, args)?;
 
                 // move buckets and proofs to this process.
                 self.receive_buckets(moving_buckets)?;
-                self.receive_proofs(moving_proofs)?;
+                self.receive_proofs(moving_proofs);
+
+                // Return borrowed snodes
+                match snode {
+                    SNodeState::Scrypto(actor, component_state) => {
+                        if let Some(component_address) = actor.component_address() {
+                            self.track.return_borrowed_global_component(
+                                component_address,
+                                component_state.unwrap(),
+                            );
+                        }
+                    }
+                    SNodeState::ResourceRef(resource_address, resource_manager) => {
+                        self.track.return_borrowed_global_resource_manager(
+                            resource_address,
+                            resource_manager,
+                        );
+                    }
+                    SNodeState::BucketRef(bucket_id, bucket) => {
+                        self.buckets.insert(bucket_id, bucket);
+                    }
+                    _ => {}
```

Comment:
A good practice is to explicitly write all the remaining matches, so we know what we're doing and not break when a new joiner adds another enum variant.

Reply:
good point